File: Haskell/HelloWorld/src/Shared/Enemy.hs
module Shared.Enemy where

import Prelude
import Data.Text (Text, pack)
import Utility (getRandomInteger)
import Shared.Constants (maxGolemLevel, golemHit, golemFlee, maxGoblinLevel, goblinHit, goblinFlee, maxSlimeLevel, slimeHit, slimeFlee)

data Enemy = Enemy
  { enemyLevel :: Integer
  , enemyHp :: Integer
  , enemyDefense :: Integer
  , enemyAttack :: Integer
  , enemyHit :: Integer 
  , enemyFlee :: Integer 
  , enemyName :: Text
  }

data EnemyType
  = Golem Enemy
  | Goblin Enemy
  | Slime Enemy

makeEnemy :: (Eq t, Num t) => t -> IO EnemyType
makeEnemy 0 = do
  level <- getRandomInteger maxGolemLevel
  pure $ Golem $ Enemy
    { enemyLevel = level
    , enemyHp = level * 50 + 100
    , enemyDefense = level * 10 + 40
    , enemyAttack = calculateGolemAttack level
    , enemyHit = golemHit 
    , enemyFlee = golemFlee 
    , enemyName = pack "ゴーレム"
    }
makeEnemy 1 = do
  level <- getRandomInteger maxGoblinLevel 
  pure $ Goblin $ Enemy
    { enemyLevel = level
    , enemyHp = level * 30 + 75
    , enemyDefense = level * 5 + 20
    , enemyAttack = level * 5 + 20
    , enemyHit = goblinHit 
    , enemyFlee = goblinFlee  
    , enemyName = pack "ゴブリン"
    }
makeEnemy 2 = do
  level <- getRandomInteger maxSlimeLevel  
  pure $ Goblin $ Enemy
    { enemyLevel = level
    , enemyHp = level * 10 + 50
    , enemyDefense = level * 2 + 10
    , enemyAttack = level * 2 + 10
    , enemyHit = slimeHit
    , enemyFlee = slimeFlee
    , enemyName = pack "スライム"
    }
makeEnemy _ = makeEnemy 0


calculateGolemAttack :: Integral a => a -> Integer
calculateGolemAttack level = toInteger $ level * 10 + 40

File: FSharp/HelloWorld/Lib/Shared/Enemy.fs
module HelloWorld.Lib.Shared.Enemy

open System

type Enemy = {
    Name: string
    mutable Hp: int
    Defense: int
    Attack: int
    Flee: int
    Hit: int
    Level: int
}

type EnemyType =
    Golem of Enemy
    | Goblin of Enemy
    | Slime of Enemy
    
let CalculateGolemAttack level = level * 10 + 40

let rec MakeEnemy ordinal =
    let rng = Random()
    match ordinal with
    | 0 ->
        let level = rng.Next(0, Constants.MaxGolemLevel)
        Golem {
            Enemy.Level = level
            Enemy.Attack = CalculateGolemAttack level
            Enemy.Defense = level * 10 + 40
            Enemy.Flee = Constants.GolemFlee
            Enemy.Hit = Constants.GolemHit
            Enemy.Name = "ゴーレム"
            Enemy.Hp = level * 50 + 100
        }
    | 1 ->
        let level = rng.Next(0, Constants.MaxGoblinLevel)
        Goblin {
            Enemy.Level = level
            Enemy.Attack = level * 5 + 20
            Enemy.Defense = level * 5 + 20
            Enemy.Flee = Constants.GoblinFlee
            Enemy.Hit = Constants.GoblinHit
            Enemy.Name = "ゴブリン"
            Enemy.Hp = level * 30 + 75
        }
    | 2 ->
        let level = rng.Next(0, Constants.MaxSlimeLevel)
        Slime {
            Enemy.Level = level
            Enemy.Attack = level * 2 + 10
            Enemy.Defense = level * 2 + 10
            Enemy.Flee = Constants.SlimeFlee
            Enemy.Hit = Constants.SlimeHit
            Enemy.Name = "スライム"
            Enemy.Hp = level * 10 + 50
        }
    | _ -> MakeEnemy 0